### DISCUSSION TITLE
Computing ECE

### DISCUSSION THREAD

**Discussion by noeliaferruz** [facebookresearch/esm#64] [Q&A]
First of all congratulations on this fantastic work and publication.

Quick question: I would like to compute the exponentiated cross-entropy (ECE) metric of these transformers and others, but not sure how to obtain the loss from the pre-computed models.

If I load ESM-1:
```
model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S()
batch_converter = alphabet.get_batch_converter()
with torch.no_grad():
    results = model(batch_tokens, repr_layers=[33], return_contacts=True)
```

How do I compute the perplexity from `results`?

Thanks

**Comment by joshim5** [comment]
Hi @noeliaferruz, thanks for the question! We compute exponentiated cross-entropy (ECE) as `exp(loss)`. To compute ECE for a batch, you should evaluate `ECE = exp(mean(loss(batch)))`, where `loss(batch)` can be computed with the standard pytorch cross-entropy. Note that this is different than the perplexity (ppl) definition given in several papers (`ppl = mean(exp(loss(batch)))`).

Let me know if that helps or you have any additional questions!

**Reply by noeliaferruz** [reply]
Thanks for the quick answer, @joshim5.

So looking at the Pytorch documentation I could obtain the token representations for a batch and then compute the loss:

```
with torch.no_grad():
    results = model(batch_tokens, repr_layers=[33], return_contacts=True)
token_representations = results["representations"][33]

loss = nn.CrossEntropyLoss()
loss = loss(token_representations, target)
```
But what is the `target` probability distribution in this case?

**Reply by joshim5** [reply]
`nn.CrossEntropyLoss` implements a categorical cross entropy loss, so `target` isn't a probability distribution, but rather a list of the targets. See [pytorch documentation here](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html).

Next, you'll want to use `results["logits"]` instead of `results["representations"][33]`. A softmax of the logits at each position represents the probabilities under the model for each amino acid at that position.

Lastly, `target` should be `batch_tokens`. Since the `batch_tokens` may contain sequences of different length, they are padded to the same length using a padding token. You should set `ignore_index` in the CrossEntropyLoss because of this.

**Reply by noeliaferruz** [reply]
Thanks for the input. I think I'm however messing up with the dimensions.
I get it working with 1 sequence but not with several. Am I supposed to loop over each sequence?

```
token_logits = results["logits"]
loss = torch.nn.CrossEntropyLoss(ignore_index=1)
loss(token_logits[0], batch_tokens[0])
tensor(0.1967)
```
```
loss(token_logits, batch_tokens)
ValueError: Expected target size (2, 33), got torch.Size([2, 73])
```
I didn't apply any softmax to token_logits because according to the documentation 'The input is expected to contain raw, unnormalized scores for each class'.  Hope this is correct.

Thanks for your help. Just getting familiar with PyTorch actually.

**Comment by eric-tc-wong** [comment]
Thanks @joshim5, for the instructions on calculating ECE, which I'm also trying to compute in order to compare performance of PLMs. I'm still having trouble reproducing the ECE from the pre-trained t6 and ESM-1b models using the validation set provided.

I calculate ece as:
loss_fn = nn.CrossEntropyLoss(reduction='none', ignore_index=-100)
...
ece = ((loss_fn(pred, noised_tokens) * noise_mask).sum() / noise_mask.sum()).exp()

I'm getting ECE > 12 for t6 model, and ECE > 9 for ESM-1b, which are a bit higher than expected.
Is the ECE calculated by masking 15% of the residues like in training or is it calculated by masking one residue per protein sequence? Masking one residue per protein results in smaller ECE that is closer to the published values. Also, is the cross-entropy calculated from 25 residues token classes instead of the 20 canonical residue classes? Any suggestions on how to get my results closer to the published figures would be appreciated. I'm only masking a random subset of positions so I won't get the exact results.

Thanks
